src/Interpreters/HashJoin.cpp

```diff
@@ -2294,14 +2318,33 @@ void HashJoin::joinBlockImplCross(Block & block, ExtraBlockPtr & not_processed)
                 const IColumn & column_right = *block_right.getByPosition(col_num).column;
                 dst_columns[num_existing_columns + col_num]->insertRangeFrom(column_right, 0, rows_right);
             }
+        };
+
+        for (const Block & compressed_block_right : data->blocks)
+        {
+            auto block_right = compressed_block_right.decompress();
+            process_right_block(block_right);
+        }
+
+        if (tmp_stream)
+        {
+            if (reader == nullptr)
+            {
+                tmp_stream->finishWritingAsyncSafe();
+                reader = tmp_stream->getReadStream();
+            }
+            while (auto block_right = reader->read())
```

Comment:
Here we will read the whole file, so we need to check termination condition `rows_added > max_joined_block_rows` and pause till next call